library/src/main/java/org/apache/apex/malhar/lib/wal/FileSystemWAL.java

```diff
@@ -0,0 +1,598 @@
+
+    private FileSystemWALWriter()
+    {
+      //for kryo
+      fileSystemWAL = null;
+    }
+
+    public FileSystemWALWriter(@NotNull FileSystemWAL<T> fileSystemWal)
+    {
+      this.fileSystemWAL = Preconditions.checkNotNull(fileSystemWal, "wal");
+    }
+
+    protected void open(@NotNull FileContext fileContext) throws IOException
+    {
+      this.fileContext = Preconditions.checkNotNull(fileContext, "file context");
+      recover();
+      if (outputStream == null) {
+        outputStream = getOutputStream(currentPointer);
+      }
+    }
+
+    private void recover() throws IOException
+    {
+      LOG.debug("current point", currentPointer);
+      String tmpFilePath = tmpFiles.get(currentPointer.getPartNum());
+      if (tmpFilePath != null) {
+
+        Path tmpPath = new Path(tmpFilePath);
+        if (fileContext.util().exists(tmpPath)) {
+          LOG.debug("tmp path exists {}", tmpPath);
+
+          outputStream = getOutputStream(currentPointer);
+          DataInputStream inputStreamOldTmp = fileContext.open(tmpPath);
+
+          IOUtils.copyPartial(inputStreamOldTmp, currentPointer.offset, outputStream);
+
+          outputStream.flush();
+          //remove old tmp
+          inputStreamOldTmp.close();
+          LOG.debug("delete tmp {}", tmpPath);
+          fileContext.delete(tmpPath, true);
+        }
+      }
+
+      //find all valid path names
+      Set<String> validPathNames = new HashSet<>();
+      for (Map.Entry<Integer, String> entry : tmpFiles.entrySet()) {
+        if (entry.getKey() <= currentPointer.partNum) {
+          validPathNames.add(new Path(entry.getValue()).getName());
+        }
+      }
+      LOG.debug("valid names {}", validPathNames);
+
+      //there can be a failure just between the flush and the actual checkpoint which can leave some stray tmp files
+      //which aren't accounted by tmp files map
+      Path walPath = new Path(fileSystemWAL.filePath);
+      Path parentWAL = walPath.getParent();
+      if (parentWAL != null && fileContext.util().exists(parentWAL)) {
+        RemoteIterator<FileStatus> remoteIterator = fileContext.listStatus(parentWAL);
+        while (remoteIterator.hasNext()) {
+          FileStatus status = remoteIterator.next();
+          String fileName = status.getPath().getName();
+          if (fileName.startsWith(walPath.getName()) && fileName.endsWith(TMP_EXTENSION) &&
+              !validPathNames.contains(fileName)) {
+            LOG.debug("delete stray tmp {}", status.getPath());
+            fileContext.delete(status.getPath(), true);
+          }
+
+        }
+      }
+
+    }
+
+    protected void close() throws IOException
+    {
+      if (outputStream != null) {
+        flush();
+        outputStream.close();
+        outputStream = null;
+        LOG.debug("closed {}", currentPointer.partNum);
+      }
+    }
+
+    @Override
+    public int append(T entry) throws IOException
+    {
+      byte[] slice = fileSystemWAL.serde.serialize(entry);
```

Comment:
Is it good to mix in serialization at this level. This requires you to make a copy when you have Slice as source.